File: 383_ransom_note/383_ransom_note.py
from typing import Dict


class Solution:
    def canConstruct(self, ransomNote: str, magazine: str) -> bool:
        reconstructedMap: Dict[str, int] = {}
        for char in magazine:
            reconstructedMap[char] = reconstructedMap.get(char, 0) + 1
        for char in ransomNote:
            reconstructedMap[char] = reconstructedMap.get(char, 0) - 1
            if reconstructedMap[char] < 0:
                return False
        return True

File: 383_ransom_note/383_ransom_note.cpp

#include <unordered_map>
#include <string>

using namespace std;
class Solution
{
public:
    bool canConstruct(string ransomNote, string magazine)
    {
        unordered_map<char, int> reconstructedMap;
        for (auto &&val : magazine)
        {
            reconstructedMap[val]++;
        }
        for (auto &&val : ransomNote)
        {
            reconstructedMap[val]--;
            if (reconstructedMap[val] < 0)
            {
                return false;
            }
        }
        return true;
    }
};